src/frontend/src/session/cursor_manager.rs

```diff
@@ -0,0 +1,272 @@
+    rw_timestamp: i64,
+) -> Result<Vec<Option<Bytes>>> {
+    // remove sqr_id, vnode ,_row_id
+    let mut new_row = vec![Some(Bytes::from(
+        convert_logstore_i64_to_epoch(rw_timestamp).to_string(),
+    ))];
+    new_row.extend(row.drain(3..row.len() - 1).collect_vec());
+    Ok(new_row)
+}
+
+pub fn build_desc(mut descs: Vec<PgFieldDescriptor>, is_snapshot: bool) -> Vec<PgFieldDescriptor> {
+    let mut new_descs = vec![
+        PgFieldDescriptor::new(
+            "rw_timestamp".to_owned(),
+            DataType::Varchar.to_oid(),
+            DataType::Varchar.type_len(),
+        ),
+        PgFieldDescriptor::new(
+            "op".to_owned(),
+            DataType::Int16.to_oid(),
+            DataType::Int16.type_len(),
+        ),
+    ];
+    if is_snapshot {
+        new_descs.extend(descs)
+    } else {
+        new_descs.extend(descs.drain(4..descs.len() - 1));
+    }
+    new_descs
+}
+
+pub enum CursorRowValue {
+    Row((Row, Vec<PgFieldDescriptor>)),
+    QueryWithNextRwTimestamp(i64, ObjectName),
+    QueryWithStartRwTimestamp(i64, ObjectName),
+}
+#[derive(Default)]
+pub struct CursorManager {
+    cursor_map: HashMap<String, (Cursor, Instant)>,
+    // Save subsciption's retentain_secs
+    cursor_retention_secs_maps: HashMap<ObjectName, Duration>,
+}
+
+impl CursorManager {
+    pub fn add_cursor_retention_secs(
+        &mut self,
+        subscription_name: ObjectName,
+        retention_secs: Duration,
+    ) {
+        self.cursor_retention_secs_maps
+            .insert(subscription_name, retention_secs);
+    }
+
+    pub fn add_cursor(&mut self, cursor: Cursor) -> Result<()> {
+        let cursor_need_drop_time = Instant::now()
+            + *self
+                .cursor_retention_secs_maps
+                .get(&cursor.subscription_name)
+                .ok_or_else(|| {
+                    ErrorCode::InternalError(format!(
+                        "Cursor can't find retention time for subscription_name: {:?}",
+                        cursor.subscription_name
+                    ))
+                })?;
+        self.cursor_map
+            .insert(cursor.cursor_name.clone(), (cursor, cursor_need_drop_time));
+        Ok(())
+    }
+
+    pub fn update_cursor(&mut self, cursor: Cursor) -> Result<()> {
+        let cursor_need_drop_time = Instant::now()
+            + *self
+                .cursor_retention_secs_maps
+                .get(&cursor.subscription_name)
+                .ok_or_else(|| {
+                    ErrorCode::InternalError(format!(
+                        "Cursor can't find retention time for subscription_name: {:?}",
+                        cursor.subscription_name
+                    ))
+                })?;
+        self.cursor_map
+            .insert(cursor.cursor_name.clone(), (cursor, cursor_need_drop_time));
+        Ok(())
+    }
+
+    pub fn remove_cursor(&mut self, cursor_name: String) -> Result<()> {
+        self.cursor_map.remove(&cursor_name);
+        Ok(())
```

Comment:
The resources occupied by the cursor include the` query iter` and `some state`, iter will be in the session empty time as well as iter timeout gc, while the state will only be in the session empty time gc, this is because the state occupies less, and the session is an easy to be gc existence, so we don't do timed clearing of the state